221 search bar

najimgh opened the pull request:
Search Asset Functionality
Assets can be searched by:  

- name 
- rul
- description
- serial number
- manufacturer
- category
- site
- location
- recommendation
- asset type name

This resolves #221

najimgh committed:
Issue (#221)
- Added live search functionality to the search bar
- Search Assets by: name, rul, description, serial number, manufacturer, category, site, location and recommendation

najimgh committed:
Issue (#221)
- Added search by asset type name
- Added comments
- Removed duplicate assetDAOImpl object

najimgh committed:
Issue (#221)
- Fixed issue where you could not sort/order based on your search result

najimgh committed:
Issue (#221)
- Added Unit test for the search bar functionality

JayChouteau commented on JavaFXUserInterface/src/main/java/controllers/AssetsController.java:
```diff
@@ -168,18 +176,33 @@ private void sortingSetUp() {
                 switch (newValue) {
                     case SORT_DEFAULT:
                         FXCollections.sort(assets, Comparator.comparingInt(Item::getId));
+                        if (!search.getText().trim().isEmpty()) {
```
Having the conditional statement every time for every switch case is a little redundant. Maybe you could use the setAssetToDisplay() method to establish a local assets list for the sort asset button to use. You would establish that local list before the switch statement. This way you could remove all the added conditional statements and keep the initial switch case with the local list of assets to display instead of the class member assets

najimgh replied:
I tried this solution at first. However, I was not able to make it work. For some reason, the list of assets would not get generated again even though the code did execute generateThumbnails() properly.

najimgh replied:
The solution we talked about did not work, I tried some other things without any success. I think the way it is now is the better option so far.

JayChouteau replied:
That is fine then. Leave it this way for now. WE will see when we implement more filters and sorting options.

JayChouteau commented on JavaFXUserInterface/src/main/java/controllers/AssetsController.java:
```diff
@@ -190,99 +213,162 @@ private void sortingSetUp() {
+            if (assetsTabPane.getSelectionModel().getSelectedItem().getId().equals("thumbnailTab")) {
+                assetsThumbPane.getChildren().clear();
+                generateThumbnails();
+            } else {
```
This is really cherry-picking, but I think that having else if with the list tab specified instead of else might provide more scalability in case there are future uses and addition of other tabs. It also clarifies for developers the intention of this conditional statement (that the following code is meant for a specific tab as well, not for all other possibilities). You can let me know what you think.

najimgh replied:
Great suggestion. The way you explained it makes more sense to use else if

JayChouteau commented on JavaFXUserInterface/src/main/java/controllers/AssetsController.java:
```diff
@@ -190,99 +213,162 @@ private void sortingSetUp() {
+        search.textProperty().addListener((observable, oldValue, newValue) -> {
+            searchedAssets.clear();
+            if (!newValue.equals("")) {
+                assets.stream().filter(o -> StringUtils.containsIgnoreCase(o.getRul().get(), newValue) || StringUtils.containsIgnoreCase(o.getName(), newValue)
```
I was simply wondering if it wouldn't have been easier and more readable to just iterate through each asset's member variables and check the StringUtils.containsIgnoreCase(...) since the calling function is used and reused. It might be more complicated as well. I am just asking.

najimgh replied:
Doing it this way or iterating through each asset's member variable would not change much except for readability and understanding of the code. Although, Iterating through the assets might result in more lines of code. 
I have added comments to explain what asset.stream().filter() does

JayChouteau replied:
Yeah, I was talking about readability and understanding of the code. But I wasn't sure what was more readable between the stream and for/if-else statements. This way is fine.

JayChouteau commented on JavaFXUserInterface/src/main/java/controllers/AssetsController.java:
```diff
@@ -190,99 +213,162 @@ private void sortingSetUp() {
+        ObservableList<Asset> assetsDisplayed = setAssetListToDisplay();
 
-        for (Asset asset : assets) {
+        if (assetsDisplayed != null) {
```
I think it would be good to provide an alternative text to display on the page in the case where there are no assets to display. What I had in mind would be kind of a grey text appearing in the middle of the screen stating that there are "No results for your search" or something like that.

najimgh replied:
"No results found" is now displayed if no match has been found from the search

JayChouteau replied:
Well done. The only comment I would make is that it feels weird that the display of "No results found" is different for the list view than the thumbnail view. Was there a specific reason for it? Is it because of the flow pane? Otherwise, I would have left it the same as the list view.

najimgh replied:
Yes, because of the flow pane I was not able to make it the same as the list view

JayChouteau commented on JavaFXUserInterface/src/main/java/controllers/AssetsController.java:
```diff
@@ -342,14 +429,16 @@ public void generateList() {
-        AnchorPane.setRightAnchor(table, 0.0);
-        AnchorPane.setLeftAnchor(table, 0.0);
-        assetsListPane.getChildren().addAll(table);
+        if (assetsDisplayed != null) {
```
I think it would be good to provide an alternative text to display on the page in the case where there are no assets to display. What I had in mind would be kind of a grey text appearing in the middle of the screen stating that there are "No results for your search" or something like that.

najimgh replied:
Table now shows "No results found"

JayChouteau replied:
Nice touch on the no results found.

JayChouteau commented on JavaFXUserInterface/src/test/java/UnitTests/Controllers/AssetsControllerTest.java:
```diff
@@ -0,0 +1,72 @@
+        release(new MouseButton[]{});
+    }
+
+    @Test
```
These tests are great tests, but they are more system tests than unit tests. At least that's how I feel, but maybe we could discuss about it at the meeting or take a few minutes to discuss it.

JayChouteau replied:
We can leave this as is, I consider this issue done

najimgh replied:
I couldn't find any relevant information for unit testing UI and more specifically the search bar

JayChouteau reviewed (changes requested):
Good work on the user story. The logic is good and all acceptance criteria have been met. Solid work overall. I left a few comments. Some of them are request changes, but others are more discussions. In any case, I think it would be great if you could answer them all.

JayChouteau commented on JavaFXUserInterface/src/test/java/UnitTests/Controllers/AssetsControllerTest.java:
```diff
@@ -0,0 +1,72 @@
+        FXMLLoader fxmlLoader = new FXMLLoader(AssetsController.class.getResource("/Assets.fxml"));
+        Parent root = fxmlLoader.load();
+        scene = new Scene(root);
+        assetController = (AssetsController) fxmlLoader.getController();
```
IntelliJ warning: the casting to (AssetsController) is redundant and be removed.

najimgh replied:
Fixed

JayChouteau commented on JavaFXUserInterface/src/test/java/UnitTests/Controllers/AssetsControllerTest.java:
```diff
@@ -0,0 +1,72 @@
+        clickOn("#search").write("This is an invalid search");
+        FlowPane rootNode = (FlowPane) scene.getRoot().lookup("#assetsThumbPane");
+        verifyThat("#search", TextInputControlMatchers.hasText("This is an invalid search"));
+        interact(() -> {
```
IntelliJ warning: This can be changed to a lambda expression.

najimgh replied:
Fixed

najimgh committed:
Issue (#221)
- Minor changes addressing change requests

TalalBazerbachi reviewed (changes requested):
I don't know if it's working as intended.
- I tried to search by recommendation 'ok' but it also included a failed asset.
<img width="1122" alt="image" src="https://user-images.githubusercontent.com/55160650/112673972-08326800-8e6e-11eb-8f41-8d7fe8434df6.png">

- When I typed '101', it gave more than one asset:
<img width="1129" alt="image" src="https://user-images.githubusercontent.com/55160650/112674094-3021cb80-8e6e-11eb-8c7f-bbcbeb14b97d.png">

-Also why is the performance of the search bar slow?

najimgh commented:
@t98b  
This is a normal behaviour of the search functionality. We compare the value of each member variable of the assets. Therefore it's normal to have multiple results for '101' since 101 can be found in RUL, Serial number, etc. To refine the search, we can order and filter the list of assets.

The performance is slow only when typing the first 2 or 3 characters as it is looking through the 704 assets available. Moreover, since this is a live search, we are generating and clearing the Thumbnails every time a character is typed. The only way I see that we could improve the performance would be to not make it a live search.

najimgh committed:
Issue (#221)
- Fixed issue 1: generateThumbnails() and generateList() would both get called every time the tab was changed no matter which tab
- Fixed issue 2: Improved performance of the search functionality by delaying the search of assets (delay of 0.25 seconds) in case the user keeps typing.
- Fixed issue 3: Showing "No results found" when searching assets and no match has been found

najimgh committed:
Issue (#221)
- Fixed issue 1: generateThumbnails() and generateList() would both get called every time the tab was changed no matter which tab
- Fixed issue 2: Improved performance of the search functionality by delaying the search of assets (delay of 0.25 seconds) in case the user keeps typing.
- Fixed issue 3: Showing "No results found" when searching assets and no match has been found

najimgh commented:
@t98b 
Improved performance of the live search functionality by delaying the search of assets (delay of 0.25 seconds) in case the user keeps typing.

JayChouteau reviewed (approved):
Nice job on improving the search performance by adding a delay in listening to characters. Also, nice job of indicating where no results were found and well done on handling null values for the sorting. The only comment I would make is that I feel like it is weird that the results not found for the list and results not found for the thumbnails are different. Other than that, pretty good job.
